File: start-museum/dotnet/Helpers/CuratorTerminal.cs
namespace MuseumExhibitStudio.Helpers;

public static class CuratorTerminal
{
    public static string AskLine(string question)
    {
        ArgumentNullException.ThrowIfNull(question);

        Console.Write(question);
        return Console.ReadLine()?.Trim() ?? string.Empty;
    }

    public static bool AskYesNo(string question, bool defaultYes)
    {
        ArgumentNullException.ThrowIfNull(question);

        var answer = AskLine($"{question} {(defaultYes ? "[Y/n]" : "[y/N]")}: ");

        if (string.IsNullOrWhiteSpace(answer))
        {
            return defaultYes;
        }

        if (answer.Equals("y", StringComparison.OrdinalIgnoreCase) ||
            answer.Equals("yes", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (answer.Equals("n", StringComparison.OrdinalIgnoreCase) ||
            answer.Equals("no", StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        return defaultYes;
    }

    public static string[] ReadFacts()
    {
        Console.WriteLine("Enter one approved fact per line. Submit a blank line when finished:");
        var facts = new List<string>();

        while (true)
        {
            var fact = AskLine(string.Empty);
            if (string.IsNullOrWhiteSpace(fact))
            {
                return facts.ToArray();
            }

            facts.Add(fact);
        }
    }

    public static void CloseTerminal()
    {
    }
}

File: start-museum/java/src/main/java/workshop/CuratorTerminal.java
package workshop;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public final class CuratorTerminal {
    private static final BufferedReader INPUT = new BufferedReader(new InputStreamReader(System.in));
    private static boolean closed;

    private CuratorTerminal() {
    }

    public static String askLine(String question) throws IOException {
        System.out.print(question);
        return readLine().trim();
    }

    public static boolean askYesNo(String question, boolean defaultYes) throws IOException {
        System.out.print(question + (defaultYes ? " [Y/n]: " : " [y/N]: "));
        String answer = readLine().trim();
        if (answer.isEmpty()) {
            return defaultYes;
        }
        if (answer.equalsIgnoreCase("y") || answer.equalsIgnoreCase("yes")) {
            return true;
        }
        if (answer.equalsIgnoreCase("n") || answer.equalsIgnoreCase("no")) {
            return false;
        }
        return defaultYes;
    }

    public static List<String> readFacts() throws IOException {
        System.out.println("Enter one approved fact per line. Submit a blank line when finished:");
        List<String> facts = new ArrayList<>();
        while (true) {
            String line = readLine();
            if (line.isBlank()) {
                break;
            }
            facts.add(line.trim());
        }
        return List.copyOf(facts);
    }

    public static void close() throws IOException {
        if (!closed) {
            closed = true;
            INPUT.close();
        }
    }

    private static String readLine() throws IOException {
        String line = INPUT.readLine();
        return line == null ? "" : line;
    }
}
